src/git.ts

```diff
@@ -0,0 +1,259 @@
+import * as promisify from 'promisify-node';
+
+// import {checkoutLatestRelease} from './latest-release';
+import * as util from './util';
+
+/**
+ * Represents GitHub repository + optional specific branch/ref requested by the
+ * tattoo user.
+ */
+export interface GitHubRepoRef {
+  // The branch name or SHA of the commit to checkout in the clone.
+  checkoutRef?: string;
+
+  // The name of the org or user who owns the repo on GitHub.
+  ownerName?: string;
+
+  // The name of the repo on GitHub.
+  repoName?: string;
+}
+
+/**
+ * GitHubConnection is a wrapper class for the GitHub npm package that
+ * assumes action as-a-user, and a minimal set of supported API calls (mostly
+ * to do with listing and cloning owned repos) using a token and building in
+ * rate-limiting functionality using the Bottleneck library to throttle API
+ * consumption.
+ */
+export class GitHubConnection {
+  private _cache: {repos: Map<string, Map<string, GitHub.Repo>>};
+  private _cloneOptions: nodegit.CloneOptions;
+  private _cloneRateLimiter: Bottleneck;
+  private _github: GitHub;
+  private _token: string;
+  private _user: GitHub.User;
+
+  constructor(token: string) {
+    this.resetCache();
+    this._token = token;
+    this._github = new GitHub({
+      version: '3.0.0',
+      protocol: 'https',
+    });
+    this._github.authenticate({type: 'oauth', token: token});
+    // TODO: Make the arguments to rate limiter configurable.
+    this._cloneRateLimiter = new Bottleneck(20, 100);
+    this._cloneOptions = {
+      fetchOpts: {
+        callbacks: {
+          certificateCheck: function() {
+            return 1;
+          },
+          credentials: function(url: string, userName: string) {
+            return nodegit.Cred.userpassPlaintextNew(token, 'x-oauth-basic');
+          }
+        }
+      }
+    };
+  }
+
+  resetCache() {
+    this._cache = {repos: new Map()};
+  }
+
+  // HACK: Don't expose repos, create the necessary methods to operate on them.
+  get repos() {
+    return this._github.repos;
+  }
+
+  // HACK: Don't expose user, create the necessary methods to operate on them.
+  get user() {
+    return this._github.user;
+  }
+
+  /**
+   * Given a github repository and a directory to clone it into, return an
+   * ElementRepo once it has been cloned and checked out.
+   */
+  async clone(githubRepo: GitHub.Repo, cloneDir: string):
+      Promise<nodegit.Repository> {
+    let nodegitRepo: nodegit.Repository;
+    if (util.existsSync(cloneDir)) {
+      let updatedRepo: nodegit.Repository;
+      nodegitRepo =
+          await nodegit.Repository.open(cloneDir)
+              .then((repo) => {
+                updatedRepo = repo;
+                return this._cloneRateLimiter.schedule(
+                    () => updatedRepo.fetchAll(this._cloneOptions.fetchOpts));
+              })
+              .then(() => updatedRepo);
+    } else {
+      // Potential race condition if multiple repos w/ the same name are
```

Comment:
I know it's from existing code, but is this comment a warning of an unhandled potential race, or a description of the handling of the race?